// @arkane-network/arkane-connect/dist/src/utils/Utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var vue_env_1 = tslib_1.__importDefault(require("../../vue.env"));
var Utils = /** @class */ (function () {
    function Utils() {
    }
    Object.defineProperty(Utils, "env", {
        get: function () {
            var env = vue_env_1.default;
            switch (Utils.environment) {
                case 'local':
                case 'tst1':
                    env.VUE_APP_REALM_PUBLIC_KEY = env.VUE_APP_REALM_PUBLIC_KEY_TST1;
                    break;
                case 'staging':
                    env.VUE_APP_REALM_PUBLIC_KEY = env.VUE_APP_REALM_PUBLIC_KEY_STAGING;
                    break;
                default:
                    env.VUE_APP_REALM_PUBLIC_KEY = env.VUE_APP_REALM_PUBLIC_KEY_PROD;
            }
            return env;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Utils, "urls", {
        get: function () {
            var prefix = '';
            switch (Utils.environment) {
                case 'local':
                    prefix = '-tst1';
                    break;
                case 'tst1':
                    prefix = '-tst1';
                    break;
                case 'staging':
                    prefix = '-staging';
                    break;
            }
            return {
                api: "https://api" + prefix + ".arkane.network/api",
                connect: Utils.environment === 'local' ? 'http://localhost:8081' : "https://connect" + prefix + ".arkane.network",
                login: "https://login" + prefix + ".arkane.network/auth",
            };
        },
        enumerable: true,
        configurable: true
    });
    Utils.getOrigin = function (url) {
        var parts = url.match(/^.+\:\/\/[^\‌​/]+/);
        return (Array.isArray(parts) && parts.length > 0) ? parts[0] : 'unknown';
    };
    Utils.shuffleArray = function (array) {
        var _a;
        for (var i = array.length - 1; i > 0; i--) {
            var j = Math.floor(Math.random() * (i + 1));
            _a = [array[j], array[i]], array[i] = _a[0], array[j] = _a[1];
        }
        return array;
    };
    Utils.removeNulls = function (obj) {
        return Object.keys(obj)
            .filter(function (key) { return obj[key] !== null && obj[key] !== undefined; }) // Remove undef. and null.
            .reduce(function (newObj, key) {
            var _a, _b;
            if (typeof obj[key] === 'object' && !Array.isArray(obj[key])) {
                return Object.assign(newObj, (_a = {}, _a[key] = Utils.removeNulls(obj[key]), _a));
            }
            else {
                return Object.assign(newObj, (_b = {}, _b[key] = obj[key], _b));
            }
        }, {});
    };
    Utils.removeNullsAndEmpty = function (obj) {
        return Object.keys(obj)
            .filter(function (key) { return obj[key] !== null && obj[key] !== undefined && obj[key] !== ''; }) // Remove undef. and null.
            .reduce(function (newObj, key) {
            var _a, _b;
            if (typeof obj[key] === 'object' && !Array.isArray(obj[key])) {
                return Object.assign(newObj, (_a = {}, _a[key] = Utils.removeNullsAndEmpty(obj[key]), _a));
            }
            else {
                return Object.assign(newObj, (_b = {}, _b[key] = obj[key], _b));
            }
        }, {});
    };
    Utils.messages = function () {
        return {
            hasValidOrigin: function (message) {
                return message.origin === Utils.urls.connect;
            },
            hasType: function (message) {
                return message.data && message.data.type && message.data.type !== '';
            },
            isOfType: function (message, eventType) {
                return Utils.messages().hasType(message) && message.data.type === eventType;
            },
        };
    };
    Utils.formatNumber = function (value, minDecimals, maxDecimals) {
        if (minDecimals === void 0) { minDecimals = 2; }
        if (maxDecimals === void 0) { maxDecimals = minDecimals; }
        return new Intl.NumberFormat('en-US', {
            minimumFractionDigits: minDecimals,
            maximumFractionDigits: maxDecimals,
        }).format(value);
    };
    Utils.rawValue = function () {
        return {
            toTokenValue: function (rawValue, decimals) { return rawValue / Math.pow(10, decimals); },
            toGwei: function (rawValue) { return rawValue / Math.pow(10, 9); },
        };
    };
    Utils.gwei = function () {
        return {
            toRawValue: function (rawValue) { return rawValue * Math.pow(10, 9); },
        };
    };
    Utils.openExternalUrl = function (url, targetBlank) {
        if (targetBlank === void 0) { targetBlank = true; }
        if (targetBlank) {
            var newWindow = window.open('', '_blank');
            if (newWindow) {
                newWindow.opener = null;
                newWindow.location.assign(url);
            }
            return newWindow;
        }
        else {
            window.location.href = url;
            return window;
        }
    };
    Utils.zeroIfUndefined = function (numberToVerify) {
        return numberToVerify ? numberToVerify : 0;
    };
    Utils.environment = 'prod';
    Utils.wallets = {
        hasWalletsForChainType: function (wallets, chain) {
            return Utils.wallets.filterWalletsForChainType(wallets, chain).length > 0;
        },
        filterWalletsForChainType: function (wallets, chain) {
            return Utils.wallets.filterWalletsForSecretType(wallets, chain.secretType);
        },
        hasWalletsForSecretType: function (wallets, secretType) {
            return Utils.wallets.filterWalletsForSecretType(wallets, secretType).length > 0;
        },
        filterWalletsForSecretType: function (wallets, secretType) {
            return wallets.filter(function (wallet) {
                return wallet.secretType === secretType;
            });
        },
    };
    return Utils;
}());
exports.default = Utils;
