// @arkane-network/arkane-connect/dist/src/models/Wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var WalletType;
(function (WalletType) {
    WalletType["THREEWAY_SHARED"] = "THREEWAY_SHARED";
})(WalletType = exports.WalletType || (exports.WalletType = {}));
var WalletAppType;
(function (WalletAppType) {
    WalletAppType["APP"] = "APP";
    WalletAppType["PERSONAL"] = "PERSONAL";
})(WalletAppType = exports.WalletAppType || (exports.WalletAppType = {}));
var WalletIntegratedApp;
(function (WalletIntegratedApp) {
    WalletIntegratedApp["FUNDREQUEST"] = "FND";
    WalletIntegratedApp["DOCK"] = "DOCK";
    WalletIntegratedApp["SAVE_HEAVEN"] = "SHA";
    WalletIntegratedApp["NONE"] = "NONE";
})(WalletIntegratedApp = exports.WalletIntegratedApp || (exports.WalletIntegratedApp = {}));
var Wallet = /** @class */ (function () {
    function Wallet() {
        this.id = '0';
        this.address = '';
        this.type = WalletAppType.PERSONAL;
        this.funds = [];
        this.app = WalletIntegratedApp.FUNDREQUEST;
        this.lastUpdated = 0;
        this.alias = '';
        this.description = '';
        this.archived = false;
        this.primary = false;
    }
    return Wallet;
}());
exports.Wallet = Wallet;
